docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,378 @@
+new port and channel identifiers:
+
+The denomination also needs to be updated when token is received on the source chain. This is done during the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` call. Below is the updated function that sets the :
+
+```golang
+// createOutgoingPacket
+// ...
+prefix := types.GetDenomPrefix(destinationPort, destinationChannel)
+source := strings.HasPrefix(amount[0].Denom, prefix)
+
+for _, denomTrace := range denomTraces {
+  // set the value to the lookup table if not stored already
+  traceHash := denomTrace.Hash()
+  if !k.HasDenomTrace(ctx, traceHash) {
+    k.SetDenomTrace(ctx, traceHash, denomTrace)
+  }
+}
+
+if source {
+  // clear the denomination from the prefix to send the coins to the escrow account
+  coins := make(sdk.Coins, len(amount))
+  for i, coin := range amount {
+    if strings.HasPrefix(denomTraces[i].Trace, prefix) {
+      if err := denomTraces[i].RemovePrefix(); err != nil {
+        return err
+      }
+
+      // set the new value to the lookup table if not stored already
+      traceHash := denomTraces[i].Hash()
+      if !k.HasDenomTrace(ctx, traceHash) {
+        k.SetDenomTrace(ctx, traceHash, denomTrace[i])
+      }
+
+      coins[i] = sdk.NewCoin(denomTrace[i].IBCDenom(), coin.Amount)
+    } else {
+      coins[i] = coin
+    }
+  }
+
+  // escrow tokens if the destination chain is the same as the sender's
+  escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
+
+  // escrow source tokens. It fails if balance insufficient.
+  if err := k.bankKeeper.SendCoins(
+    ctx, sender, escrowAddress, coins,
+  ); err != nil {
+    return err
+  }
+} else {
+  // build the receiving denomination prefix if it's not present
+  prefix = types.GetDenomPrefix(sourcePort, sourceChannel)
+  for _, coin := range amount {
+    if !strings.HasPrefix(denomTrace[i].Trace, prefix) {
+      return Wrapf(types.ErrInvalidDenomForTransfer, "invalid token denom trace prefix: %s", denomTrace[i].IBCDenom())
+    }
+  }
+// ...
+}
+```
+
+```golang
+// OnRecvPacket
+// ...
+prefix := types.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
+source := strings.HasPrefix(data.DenomTraces[0].Trace, prefix)
+
+// ...
+
+for _, denomTrace := range denomTraces {
+  // set the value to the lookup table if not stored already
+  traceHash := denomTrace.Hash()
+  if !k.HasDenomTrace(ctx, traceHash) {
+    k.SetDenomTrace(ctx, traceHash, denomTrace)
+  }
+}
+
+if source {
+  // ... (no changes)
+}
+
+// check the denom prefix
+prefix = types.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+coins := make(sdk.Coins, len(data.Amount))
+for i, coin := range data.Amount {
+  if !strings.HasPrefix(denomTraces[i].Trace, prefix) {
+    return Wrapf(types.ErrInvalidDenomForTransfer, "invalid token denom trace prefix: %s", denomTrace[i].IBCDenom(),
+    )
+  }
+
+  if err := denomTraces[i].RemovePrefix(); err != nil {
```

Comment:
Shouldn't we be removing prefix if source == true?

Reply:
hmm not according to the implementation